### ISSUE TITLE
Allow Bedrock Converse to accept base64

### ISSUE THREAD

**Issue by sterankin** [aws/aws-sdk-js-v3#6703]
### Describe the feature

When calling a LLM (such as Claude or Llama3) using the Converse Command in the SDK, it specifies that the image must be in a Uint8Array:

https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/bedrock-runtime/command/ConverseCommand/

It is my understanding that the SDK will then use the smithy library to convert the image to base64.  It would be preferable if the SDK would also accept base64 in addition to Uint8Array.

https://github.com/aws/aws-sdk-js-v3/tree/main/clients/client-bedrock-runtime

### Use Case

The use case is simple.  Bedrock LLMs are often exposed using API Gateway and proxy lambdas.  Clients will interact with the API using REST calls.

It is not possible to send an image as a Uint8Array via a REST call, however it is possible to base64 encode the image and send.

### Proposed Solution

I suggest that the Converse Command should accept base64 in the bytes field and send it to the model as is, it should only convert the bytes if its a Uint8Array.  I think[ this file](https://github.com/aws/aws-sdk-js-v3/blob/main/clients/client-bedrock-runtime/src/protocols/Aws_restJson1.ts#L924) needs to change.

```
content: [ // ContentBlocks // required
        { // ContentBlock Union: only one key present
          text: "STRING_VALUE",
          image: { // ImageBlock
            format: "png" || "jpeg" || "gif" || "webp", // required
            source: { // ImageSource Union: only one key present
              **bytes**: **base64** string _OR_ new Uint8Array(), 
            },
          },
}
]
```

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### SDK version used

3.682.0

### Environment details (OS name and version, etc.)

Node

**Comment by zshzbh** [comment]
Hey @sterankin ,

The service api accepts base64 for image([ref](https://docs.aws.amazon.com/bedrock/latest/APIReference/API_runtime_ImageSource.html)) but SDK provides the ability to use `Uint8Array`. 

If you are not using SDK, you can use image of `base64` type to send a `HTTPRequest`. 

Or if you are using AWS JS SDK, SDK could handle the type conversation and send image with `Uint8Array`.

After knowing the info, do you still want to add `base64` support for `image`?

**Comment by sterankin** [comment]
@zshzbh Thanks for your reply.  I understand that the service api accepts base64, however a common pattern to access the models will be via API Gateway and Lambda (using SDK).

You cannot send Uint8Array via a REST API call to an API Gateway, it needs to be sent as `base64`.  Therefore, the SDK should accept and use the base64 data when calling the model.  It should accept `base64` _and_ `Uint8Array`.  This remove the step of having to convert the `base64` to `Uint8array` first, only for the SDK to then convert it back to `base64`.

In summary, the sdk should be changed to check if the format of the image is already `base64`, and if so send it to the model as is.  If the image format is `Uint8Array` then the behaviour can continue as it is currently.

**Comment by kuhe** [comment]
We cannot change the behavior of the AWS SDK in this case. The default base64 encoder already converts strings from utf8 to byte and we do not want to add an additional check to see if the string is in the base64 format.

Another reason is that the wire format of an AWS SDK client is intended to be opaque to the user. Even though it involves JSON at this point, it may change at a later date into a format that prefers binary (I am not saying that it will). 

You can configure this on your client instance.

```js
import { toBase64 } from "@smithy/util-base64";

new BedrockRuntimeClient({
  base64Encoder: (bytes: string | Uint8Array) => {
    if (typeof bytes === "string") return bytes;
    return toBase64(bytes);
  }
});
```

In this example, the client's base64 encoder is overridden to assume strings are already in base64 format. This is even compatible with the unlikely scenario that the wire format changes, since in that event the base64 encoder will not be called.

**Comment by sterankin** [comment]
> We cannot change the behavior of the AWS SDK in this case. The default base64 encoder already converts strings from utf8 to byte-array and we do not want to add an additional check to see if the string is in the base64 format.
> 
> Another reason is that the wire format of an AWS SDK client is intended to be opaque to the user. Even though it involves JSON at this point, it may change at a later date into a format that prefers binary (I am not saying that it will).
> 
> You can configure this on your client instance.
> 
> ```ts
> import { toBase64 } from "@smithy/util-base64";
> 
> new BedrockRuntimeClient({
>   base64Encoder: (bytes: string | Uint8Array) => {
>     if (typeof bytes === "string") return bytes;
>     return toBase64(bytes);
>   }
> });
> ```
> 
> In this example, the client's base64 encoder is overridden to assume strings are already in base64 format. This is even compatible with the unlikely scenario that the wire format changes, since in that event the base64 encoder will not be called.
> 
> If you use TypeScript, you may need to override a few more things to get this polymorphic encoder to work. It is possible to apply a type transform on the client to accept strings where Uint8Array is usually required. We don't have anything out of the box for this, but the concept can be seen here: https://github.com/smithy-lang/smithy-typescript/blob/main/packages/types/src/transform/type-transform.ts#L42 in `type ConditionalRecursiveTransformExact<T, FromType, ToType>`.

Thanks for your response.  Just to confirm - the code snippet you provided would remove the need for me to convert the base64 image to Uint8Array (sent to the lambda via API Gateway).

Instead it would allow the Bedrockruntime client to accept base64 when I pass in the converse command?

This is what I am having to do in the lambda:

`    // check if there is an image in the request and convert it to uint8array from base64
    requestBody?.messages?.forEach((msg) => {
      msg?.content?.forEach((ct) => {
        if (ct.image && ct.image.source && 'bytes' in ct.image.source) {
          // Replace the bytes value
          console.log('replacing image by converting to uint8array');
          const sdkImageBytes = Buffer.from(ct.image.source.bytes as unknown as string, 'base64');
          const uint8Array = new Uint8Array(sdkImageBytes);
          ct.image.source.bytes = uint8Array;
        }
      });
    });

    command = new ConverseCommand({
      modelId,
      messages: requestBody ? requestBody.messages : [],
      inferenceConfig: requestBody?.inferenceConfig || {},
      system: requestBody?.system || [],
      ...guardrailsParams,
    });
  }

    const response = await client.send(command);
`

with your code can I just send requestBody.messages in base64?

Thanks

**Comment by kuhe** [comment]
Yes.

The role of the `base64Encoder` client configuration option is to convert byte arrays into base64 strings for formats (like JSON) that do not support byte arrays. This transform function is only called for input fields that are designated as bytes, for example [ImageSource.bytes](https://docs.aws.amazon.com/bedrock/latest/APIReference/API_runtime_ImageSource.html).

With the override function, if it sees an input of type string, no conversion is performed. This assumes the input string is base64.

**Comment by sterankin** [comment]
Thanks for the help here.  Closing this issue

**Comment by sterankin** [comment]
I tried the code in my lambda:

```
  const client = new BedrockRuntimeClient({
    region: 'us-east-1',
    // this ensures we dont have to convert if its already base64
    base64Encoder: (bytes: string | Uint8Array) => {
      if (typeof bytes === 'string') return bytes;
      return toBase64(bytes);
    },
  });

    command = new ConverseCommand({
      modelId,
      messages: requestBody ? requestBody.messages : [],
      inferenceConfig: requestBody?.inferenceConfig || {},
      system: requestBody?.system || [],
      ...guardrailsParams,
    });

const response = await client.send(command);
```

But it returns a 400 SerializationException: UnknownError when I send an image as a base64 encoded string

**Comment by kuhe** [comment]
Examine the outgoing request when you use the SDK's default base64encoder (input a byte array), and compare with your base64Encoder.

It could be a base64 padding issue?
